atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java

```diff
@@ -0,0 +1,255 @@
+        }
+    }
+
+    @Test
+    public void processStartTimestampUpdateOnMultipleBatches() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        verifyStage();
+
+        Set<Long> secondTimestamps = ImmutableSet.of(5L, 123L);
+        eventCache.processStartTransactionsUpdate(secondTimestamps, SUCCESS);
+        verifyStage();
+    }
+
+    @Test
+    public void getCommitUpdateDoesNotContainCommitLocks() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processGetCommitTimestampsUpdate(COMMIT_UPDATE, SUCCESS);
+        verifyStage();
+
+        CommitUpdate commitUpdate = eventCache.getCommitUpdate(1L);
+        assertThat(commitUpdate.accept(new CommitUpdateVisitor()))
+                .containsExactlyInAnyOrder(DESCRIPTOR);
+    }
+
+    @Test
+    public void cacheClearedOnSnapshotUpdate() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, SUCCESS);
+        verifyStage();
+
+        LockWatchStateUpdate snapshot2 = LockWatchStateUpdate.snapshot(LEADER, 7L, ImmutableSet.of(DESCRIPTOR),
+                ImmutableSet.of());
+        Set<Long> timestamps3 = ImmutableSet.of(123L, 1255L);
+        eventCache.processStartTransactionsUpdate(timestamps3, snapshot2);
+        verifyStage();
+    }
+
+    @Test
+    public void getEventsForTransactionsReturnsSnapshotWithOldEvents() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processGetCommitTimestampsUpdate(COMMIT_UPDATE, SUCCESS);
+        eventCache.removeTransactionStateFromCache(START_TS);
+        verifyStage();
+
+        LockWatchStateUpdate success2 = LockWatchStateUpdate.success(LEADER, 10L, ImmutableList.of(LOCK_EVENT_2));
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, success2);
+        verifyStage();
+
+        TransactionsLockWatchUpdate results = eventCache.getUpdateForTransactions(TIMESTAMPS_2, Optional.empty());
+        assertThat(results.clearCache()).isTrue();
+        assertThat(results.startTsToSequence()).containsExactlyInAnyOrderEntriesOf(
+                ImmutableMap.of(16L, IdentifiedVersion.of(LEADER, 10L)));
+        assertThat(results.events()).containsExactly(
+                LockWatchCreatedEvent.builder(ImmutableSet.of(REFERENCE),
+                        ImmutableSet.of(DESCRIPTOR, DESCRIPTOR_3)).build(6L),
+                LOCK_EVENT_2);
+    }
+
+    @Test
+    public void failedUpdateClearsAllCaches() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, LockWatchStateUpdate.failed(LEADER));
+        verifyStage();
+    }
+
+    @Test
+    public void leaderChangeClearsCaches() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2,
```

Comment:
Actually shouldn't this throw? You have received a success update from timelock by you don't know what the snapshot is.

Reply:
Ah, I see you handle this. I would just say that this is probably an unexpected state to be in, but maybe possible if leaders switch quickly somehow, and given concurrency between startTransaction and commitTimestamp calls.

Reply:
yeah - you can imagine a case where the success update comes before the snapshot somehow, so in that case we just clear the cache outright, and wait for the snapshot update.